pkg/client/proxy_backup.go

```diff
@@ -0,0 +1,342 @@
+package client
+
+import (
+	"encoding/json"
+	"fmt"
+	"strings"
+
+	"github.com/golang/protobuf/ptypes"
+	"github.com/pkg/errors"
+	"github.com/sirupsen/logrus"
+
+	"github.com/longhorn/backupstore"
+
+	rpc "github.com/longhorn/longhorn-instance-manager/pkg/imrpc"
+)
+
+func (c *ProxyClient) SnapshotBackup(serviceAddress,
+	backupName, snapshotName, backupTarget,
+	backingImageName, backingImageChecksum string,
+	labels map[string]string, envs []string) (backupID, replicaAddress string, err error) {
+	if serviceAddress == "" {
+		return "", "", errors.Wrapf(ErrParameter, "failed to backup snapshot")
+	}
+
+	log := logrus.WithFields(logrus.Fields{"serviceURL": c.ServiceURL})
+	log.Debugf("Backing up snapshot %v to %v via proxy", snapshotName, backupName)
+
+	req := &rpc.EngineSnapshotBackupRequest{
+		ProxyEngineRequest: &rpc.ProxyEngineRequest{
+			Address: serviceAddress,
+		},
+		Envs:                 envs,
+		BackupName:           backupName,
+		SnapshotName:         snapshotName,
+		BackupTarget:         backupTarget,
+		BackingImageName:     backingImageName,
+		BackingImageChecksum: backingImageChecksum,
+		Labels:               labels,
+	}
+	recv, err := c.service.SnapshotBackup(c.ctx, req)
+	if err != nil {
+		return "", "", errors.Wrapf(err, "failed to backup snapshot %v to %v via proxy %v to %v", snapshotName, backupName, c.ServiceURL, serviceAddress)
+	}
+
+	return recv.BackupId, recv.Replica, nil
+}
+
+func (c *ProxyClient) SnapshotBackupStatus(serviceAddress, backupName, replicaAddress string) (status *SnapshotBackupStatus, err error) {
```

Comment:
This ensures the returning object is able to use the fallback clients.

Reply:
Considering the backward compatibility, I didn't directly use the proto-generated struct outside of the gRPC client/server since it may change version by version.